dts/bindings/base/mpu-region.yaml

```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2021, Carlo Caione <ccaione@baylibre.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: MPU region
+
+compatible: "mpu-region"
+
+include: [base.yaml, mem-region.yaml]
+
+properties:
+  zephyr,memory-region:
+    required: true
+
+  zephyr,memory-region-mpu:
```

Comment:
What is the reasoning behind adding this as a new compatible instead of adding the `zephyr,memory-region-mpu` property to `mem-region.yaml`? Is it solely to enable the usage of `DT_FOREACH_STATUS_OKAY_VARGS` to iterate over usages of `zephyr,memory-region-mpu`?

If so I wonder if it would make more sense to add a DT macro for iterating over all nodes that share a given property (Maybe limited to the `zephyr,` prefix).
This would also enable automatic generation of the linker regions as well:
https://github.com/zephyrproject-rtos/zephyr/blob/449c37808a68824faa93b77fe24f39f05e9a9939/include/arch/arm/aarch32/cortex_m/scripts/linker.ld#L85-L90

Reply:
> What is the reasoning behind adding this as a new compatible instead of adding the `zephyr,memory-region-mpu` property to `mem-region.yaml`? Is it solely to enable the usage of `DT_FOREACH_STATUS_OKAY_VARGS` to iterate over usages of `zephyr,memory-region-mpu`?

Yes. In the first version of this PR it was done like that but that was changed when the usage of the additional compatible was proposed, see https://github.com/zephyrproject-rtos/zephyr/pull/41223#discussion_r769749839

Reply:
> Yes. In the first version of this PR it was done like that but that was changed when the usage of the additional compatible was proposed, see #41223 (comment)

Right, that's where the compatible was proposed, but there was no discussion of its merits vs alternatives.
The iteration you really want is over anything that sets `zephyr,memory-region-mpu`, with the compatible being a bit of a hack to get that. The code I linked wants to do a similar thing eventually, iterate over all `zephyr,memory-region`.

Given that both systems want to do the same thing, their solution should probably be the same. A new compatible could be added to every SRAM node so that it does the same thing as this PR, or we can introduce a way to do `DT_FOREACH_NODE_WITH_PROP(zephyr_memory_...)` directly.

Reply:
@mbolivar-nordic what's your opinion on this?

@JordanYates the problem i see with the `DT_FOREACH_NODE_WITH_PROP()` macro is that you have no control over the properties of the nodes you are cycling through. At least when using an additional compatible you can use a `yaml` file to exactly define what properties you expect (or you need) to find and use.

Reply:
I'll need some more time to look at this, but I'm firing off a quick response in case it's helpful.

The zephyr DT tooling supports multiple compatibles in a single node, and unless I am mistaken, DT_FOREACH_STATUS_OKAY will work on a node that has multiple compatibles, even for compatibles that aren't the "main" compatible that matches a binding.

So for example if you have

```
foo { compatible = "a", "x"; };
bar { compatible = "b", "x"; };
```

Then you should be able to do `DT_FOREACH_STATUS_OKAY(x, my_fn)` even if there isn't a binding for `x`.

So maybe you want to have a 'marker' compatible that you can put on the nodes you are interested in, even if they have their own individual bindings?

If that's tedious or nonsense, feel free to ignore

Reply:
I opened #42008 to move to the multiple compatibles solution more gradually.